From the Debian firefox package:
Add more platforms having little endian byte order.

--- fdlibm/fdlibm.h.orig	2006-02-24 10:27:19.358388000 +0100
+++ fdlibm/fdlibm.h	2006-05-02 19:18:21.081317420 +0200
@@ -61,7 +61,7 @@
 #define __LITTLE_ENDIAN
 #endif
 
-#if defined(linux) && defined(__i386__)
+#if defined(linux) && (defined(__i386__) || defined(__x86_64__) || defined(__ia64) || (defined(__mips) && defined(__MIPSEL__)))
 #define __LITTLE_ENDIAN
 #endif
 
