--- ConfWindow.py.orig	Wed May 29 23:04:27 2002
+++ ConfWindow.py	Wed May 29 23:05:07 2002
@@ -526,29 +526,25 @@
 
 
     def onCmdProxyStart (self, sender, sel, ptr):
-        from wc import daemon
-        daemon.start(parent_exit=0)
+        os.system("/etc/init.d/webcleaner start")
         #debug(BRING_IT_ON, "webcleaner start")
         return 1
 
 
     def onCmdProxyStop (self, sender, sel, ptr):
-        from wc import daemon
-        daemon.stop()
+        os.system("/etc/init.d/webcleaner stop")
         #debug(BRING_IT_ON, "webcleaner stop")
         return 1
 
 
     def onCmdProxyRestart (self, sender, sel, ptr):
-        from wc import daemon
-        daemon.restart(parent_exit=0)
+        os.system("/etc/init.d/webcleaner restart")
         #debug(BRING_IT_ON, "webcleaner restart")
         return 1
 
 
     def onCmdProxyReload (self, sender, sel, ptr):
-        from wc import daemon
-        daemon.reload()
+        os.system("/etc/init.d/webcleaner reload")
         #debug(BRING_IT_ON, "webcleaner reload")
         return 1
 
